glambie/data/read_and_edit_bucket_csv_files.py

```diff
@@ -0,0 +1,343 @@
+
+    # First, check for any non equal end_dates and subsequent start_dates
+    start_dates = submission_data_frame.start_date.values
+    end_dates = submission_data_frame.end_date.values
+    start_dates = np.append(start_dates, np.nan)  # Add an extra element to the end of this list for check below
+
+    date_check_bool = True
+    nodata_check_bool = True
+
+    for ind, end_date in enumerate(end_dates):
+        # If end_date != start_date for any of the consecutive rows, set date_check_bool to False for this file - i.e.
+        # file does not pass test and will need to be edited
+        row_bool = end_date.__eq__(start_dates[ind + 1])
+        if not row_bool:
+            date_check_bool = False
+
+    # Next, check for any instances of extreme values of change - we are aware of instances where people have used
+    # e.g change = 9999. for rows where they don't have a measured change. They were unaware that they should just
+    # remove these rows instead of setting an arbitrary nodata value. Extreme value threshold needs to be different
+    # depending on the units used.
+    change_values = submission_data_frame.glacier_change_observed.values
+
+    if 'm' in submission_data_frame.unit.values[0]:  # check if units are m or mwe
+        nodata_check_bool = all(abs(i) < 100 for i in change_values)  # check that all changes in list are < +/-100
+    elif 'Gt' in submission_data_frame.unit.values[0]:  # check if units are Gt
+        nodata_check_bool = all(abs(i) < 10000 for i in change_values)
+
+    # If all rows passed the date check above, we store date_check_satisfied = True for this file: don't need to edit it
+    file_check_dataframe.loc[file_check_dataframe.local_filepath.__eq__(csv_file_path),
+                             'date_check_satisfied'] = date_check_bool
+
+    # Likewise if all rows passed the nodata check above, we store nodata_check_satisfied for this file.
+    file_check_dataframe.loc[file_check_dataframe.local_filepath.__eq__(csv_file_path),
+                             'nodata_check_satisfied'] = nodata_check_bool
+
+    return file_check_dataframe
+
+
+def edit_local_copies_of_glambie_csvs(file_check_dataframe: pd.DataFrame, local_path: str) -> pd.DataFrame:
+    """
+    Function to edit the local copies of glambie csvs that didn't pass the checks run by
+    check_glambie_submission_for_errors. Save out updated copies to the same local filepath.
+
+    Parameters
+    ----------
+    file_check_dataframe : pd.DataFrame
+        Dataframe containing results of checks for inconsistencies within the csv files compared to the expected glambie
+        standard.
+    """
+    archive_path = os.path.join(local_path, 'original_files_pre_edits')
+    os.makedirs(archive_path, exist_ok=True)
+
+    file_check_dataframe['reason_for_edit'] = ['-' for i in range(len(file_check_dataframe))]
+
+    for _, file in file_check_dataframe.iterrows():
+        if not file.date_check_satisfied:
```

Comment:
This large nested if between row 198 and 320 is quite hard to follow. Consider breaking up somehow. e.g fix_dates, fix_nodata. Or perhaps extracting the non-grace interpolation into its own private function would help.

Reply:
Have had a go at simplifying this, by extracting some large chunks into separate functions.

Reply:
Thanks - that's much easier to digest (and test!)